python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -451,6 +447,53 @@ def put(self, inMemoryDataset: Any, ref: DatasetRef) -> None:
         if self._transaction is not None:
             self._transaction.registerUndo("put", self.remove, ref)
 
+    def put_new(self, inMemoryDataset: Any, ref: DatasetRef) -> Mapping[str, DatasetRef]:
+        # Docstring inherited from base class.
+        log.debug("Put %s", ref)
+
+        # Confirm that we can accept this dataset
+        if not self.constraints.isAcceptable(ref):
+            # Raise rather than use boolean return value.
+            raise DatasetTypeNotSupportedError(
+                f"Dataset {ref} has been rejected by this datastore via configuration."
+            )
+
+        isPermanent = False
+        nsuccess = 0
+        npermanent = 0
+        nephemeral = 0
+        stored_refs: dict[str, DatasetRef] = {}
+        for datastore, constraints in zip(self.datastores, self.datastoreConstraints):
+            if (
+                constraints is not None and not constraints.isAcceptable(ref)
+            ) or not datastore.constraints.isAcceptable(ref):
+                log.debug("Datastore %s skipping put via configuration for ref %s", datastore.name, ref)
+                continue
+
+            if datastore.isEphemeral:
+                nephemeral += 1
+            else:
+                npermanent += 1
+            try:
+                stored_ref_map = datastore.put_new(inMemoryDataset, ref)
+                stored_refs.update(stored_ref_map)
+                nsuccess += 1
+                if not datastore.isEphemeral:
+                    isPermanent = True
+            except DatasetTypeNotSupportedError:
+                pass
+
+        if nsuccess == 0:
+            raise DatasetTypeNotSupportedError(f"None of the chained datastores supported ref {ref}")
+
+        if not isPermanent and npermanent > 0:
+            warnings.warn(f"Put of {ref} only succeeded in ephemeral databases", stacklevel=2)
+
+        if self._transaction is not None:
+            self._transaction.registerUndo("put", self.remove, ref)
```

Comment:
This is a side issue, but `DatastoreTransaction` is definitely something I'd hoped we could remove after the DMTN-249 transition is complete and we require file artifacts to be successfully written or transferred _before_ we insert the corresponding rows into registry.

I'm hoping this is still here just because this PR doesn't try to get us all the way there?

Reply:
Yes, this PR is mostly to get a sense that new `DatasetRef` with all extra stuff in it looks reasonable, there are many things yet to update, including transaction handling by Datastore.